### ISSUE TITLE
Regularizing predictions from variational gp

### ISSUE THREAD

**Issue by srn284** [cornellius-gp/gpytorch#1595]
Hello all, 
I am trying to implement the following objective function:
<img width="559" alt="Screenshot 2021-04-27 at 9 51 13 am" src="https://user-images.githubusercontent.com/15525940/116213936-207ff600-a73e-11eb-9e95-b98058dc46aa.png">

I am using a neural network as the feature extractor, and I am using the Multitask variational GP for the two tasks (Q(t=1,x) and Q(t=,x)). Additionally, the feature extractor has a neural network (linear) prediction head that predicts _g(x)_ (the label is t_i in the image). And an epsilon parameter i train separately.

I implemented this regularization (Targreg) in the likelihood itself. but it does very badly.

```
class MultitaskBernoulliLikelihood(Likelihood):
    def forward(self, function_samples, **kwargs):
        output_probs = torch.distributions.Normal(0, 1).cdf(function_samples)
        return Independent(torch.distributions.Bernoulli(probs=output_probs), 1)

    def expected_log_prob(self, observations, function_dist, *args, **kwargs):
        mask = kwargs['mask']
        tpred = kwargs['tpred']
        # tpred is g(x)
        tlabel = kwargs['tlabel']
        ylabel = kwargs['ylabel']
        eps = kwargs['eps']

        likelihood_samples = self._draw_likelihood_samples(function_dist, *args, **kwargs)
        res = likelihood_samples.log_prob(observations, mask)
        # custom masked log prob
        res = res.mean(dim=0)

        ytrue = ylabel
        ttrue = tlabel

        likelihood_samples = torch.mean(likelihood_samples.mean, 0).transpose(1, 0)
        # i use the likelihood samples for the computation of the regularizer below. can i do this?
        y0 = likelihood_samples[0]
        y1 = likelihood_samples[1]
        #         print(likelihood_samples)
        tpred = (tpred + 0.01) / 1.02

        ypred = ttrue * y1 + (1 - ttrue) * y0
        h = (ttrue / tpred) - (1 - ttrue) / (1 - tpred)
        ypert = ypred + eps * h


        targreg = torch.sum(torch.square(ytrue - ypert))
        # targreg is the regularization in the picture above

        return res, targreg
```

This works pretty badly i have seen in the results. I have instead tried including regularization in the training loop instead as follows:
```
log_lik, kl_div, log_prior = mll(out, outLabel, oneHotT)
# custom MLL with masking above
tpred = treatOut[:,1]
# tpred is g(x)
epsilonOut = model.Eps(tpred)
ytrue = outLabel.squeeze(-1)
ttrue = treatLabel.squeeze(-1)
outFull = torch.mean(likelihood(out).mean, 0).transpose(1,0).unsqueeze(-1)

y0 = outFull[0]
y1 = outFull[1]

tpred = (tpred+0.01)/1.02

ypred = ttrue *y1 + (1-ttrue)*y0
h=(ttrue/tpred )- ( 1-ttrue)/(1-tpred)
ypert = ypred+epsilonOut*h
targreg = torch.mean(torch.square(ytrue - ypert))
# targreg is the loss function in the image above


outcomeLoss = -(log_lik - kl_div + log_prior)

totalL = 1 * (lossT)+outcomeLoss + targreg
# totalL is lossT (the g(x) and t predictive loss) and the outcome loss (mll) and the targreg loss defined in the image 


      
```

This also does very badly in results, so I am trying to see if I should have done something else when referencing the g(x) or any of the other variables?

Perhaps is there a better or correct way :
1) to use the outputs of the multitask GP for another loss function besides the MLL?
2) to use the g(x) from the linear neural network prediction layer in the regularization better? not sure.

Thank you very much for the help!

Warm wishes,
Shishir
